### ISSUE TITLE
Incorrect BATCAP values from serial output

### ISSUE THREAD

**Issue by ProtoFoo** [rcdrones/UPSPACK_V2#4]
I implemented a Python shutdown service based on the example code in this repository. Sometimes a low battery state was reported unexpectedly and the system would shut down.

Because of this, I started logging changes to the BATCAP value. It seems that sometimes wrong values are sent by the UPS. After the dash is the raw serial string from the UPS. The leading value of BATCAP sometimes gets cut off:

```
[2020-04-29 00:34:04] Battery capacity  82 % - $ SmartUPS V1.00,Vin NG,BATCAP 82,Vout 5136 $
[2020-04-29 00:34:33] Battery capacity  81 % - $ SmartUPS V1.00,Vin NG,BATCAP 81,Vout 5136 $
[2020-04-29 00:34:49] Battery capacity   8 % - $ SmartUPS V1.00,Vin NG,BATCAP 8,Vout 5136 $
[2020-04-29 00:34:50] Battery capacity  81 % - $ SmartUPS V1.00,Vin NG,BATCAP 81,Vout 5136 $
[2020-04-29 00:35:04] Battery capacity  80 % - $ SmartUPS V1.00,Vin NG,BATCAP 80,Vout 5136 $

```
```
[2020-04-29 01:07:49] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5146 $
[2020-04-29 02:45:51] Battery capacity  10 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 10,Vout 5144 $
[2020-04-29 02:45:52] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 04:04:52] Battery capacity   0 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144 $
[2020-04-29 04:04:53] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 06:59:24] Battery capacity  10 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 10,Vout 5144 $
[2020-04-29 06:59:25] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
```
```
[2020-04-29 10:02:31] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 10:23:57] Battery capacity   0 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144 $
[2020-04-29 10:23:58] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5143 $
[2020-04-29 10:36:57] Battery capacity   0 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144 $
[2020-04-29 10:36:58] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 15:20:31] Battery capacity  10 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 10,Vout 5144 $
[2020-04-29 15:20:32] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
```
```
[2020-04-30 15:46:52] Battery capacity  72 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 72,Vout 5138 $
[2020-04-30 15:52:39] Battery capacity   7 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 7,Vout 5138 $
[2020-04-30 15:52:40] Battery capacity  72 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 72,Vout 5138 $

```

As you can see, in a line like `SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144`  there is clearly a character cut off from the BATCAP value.

This might be a bug in the firmware on the Nuvoton N76E003 AT20 microcontroller. Can you provide a fixed firmware for flashing or maybe the source code for the board, so I can fix it myself?

Thank you!

**Comment by ashbashmash** [comment]
Do the dodgy lines all miss the trailing percent sign? Not a great fix, but maybe if you only accept values when that % is there, and force a re-read if not?On 3 May 2020 22:41, ProtoFoo <notifications@github.com> wrote:
I implemented a Python shutdown service based on the example code in this repository. Sometimes a low battery state was reported unexpectedly and the system would shut down.
Because of this, I started logging changes to the BATCAP value. It seems that sometimes wrong values are sent by the UPS. After the dash is the raw serial string from the UPS. The leading value of BATCAP sometimes gets cut off:
[2020-04-29 00:34:04] Battery capacity  82 % - $ SmartUPS V1.00,Vin NG,BATCAP 82,Vout 5136 $
[2020-04-29 00:34:33] Battery capacity  81 % - $ SmartUPS V1.00,Vin NG,BATCAP 81,Vout 5136 $
[2020-04-29 00:34:49] Battery capacity   8 % - $ SmartUPS V1.00,Vin NG,BATCAP 8,Vout 5136 $
[2020-04-29 00:34:50] Battery capacity  81 % - $ SmartUPS V1.00,Vin NG,BATCAP 81,Vout 5136 $
[2020-04-29 00:35:04] Battery capacity  80 % - $ SmartUPS V1.00,Vin NG,BATCAP 80,Vout 5136 $


[2020-04-29 01:07:49] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5146 $
[2020-04-29 02:45:51] Battery capacity  10 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 10,Vout 5144 $
[2020-04-29 02:45:52] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 04:04:52] Battery capacity   0 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144 $
[2020-04-29 04:04:53] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 06:59:24] Battery capacity  10 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 10,Vout 5144 $
[2020-04-29 06:59:25] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $

[2020-04-29 10:02:31] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 10:23:57] Battery capacity   0 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144 $
[2020-04-29 10:23:58] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5143 $
[2020-04-29 10:36:57] Battery capacity   0 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144 $
[2020-04-29 10:36:58] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 15:20:31] Battery capacity  10 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 10,Vout 5144 $
[2020-04-29 15:20:32] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $

[2020-04-30 15:46:52] Battery capacity  72 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 72,Vout 5138 $
[2020-04-30 15:52:39] Battery capacity   7 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 7,Vout 5138 $
[2020-04-30 15:52:40] Battery capacity  72 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 72,Vout 5138 $


As you can see, in a line like SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144  there is clearly a character cut off from the BATCAP value.
This might be a bug in the firmware on the Nuvoton N76E003 AT20 microcontroller. Can you provide a fixed firmware for flashing or maybe the source code for the board, so I can fix it myself?
Thank you!

**Comment by ashbashmash** [comment]
Scrap that, I was looking at the wrong lines! Maybe store the last recorded battery percent, and if it's widly different assume it's string and read again?On 5 May 2020 09:34, ash@lolosmith.com wrote:Do the dodgy lines all miss the trailing percent sign? Not a great fix, but maybe if you only accept values when that % is there, and force a re-read if not?On 3 May 2020 22:41, ProtoFoo <notifications@github.com> wrote:
I implemented a Python shutdown service based on the example code in this repository. Sometimes a low battery state was reported unexpectedly and the system would shut down.
Because of this, I started logging changes to the BATCAP value. It seems that sometimes wrong values are sent by the UPS. After the dash is the raw serial string from the UPS. The leading value of BATCAP sometimes gets cut off:
[2020-04-29 00:34:04] Battery capacity  82 % - $ SmartUPS V1.00,Vin NG,BATCAP 82,Vout 5136 $
[2020-04-29 00:34:33] Battery capacity  81 % - $ SmartUPS V1.00,Vin NG,BATCAP 81,Vout 5136 $
[2020-04-29 00:34:49] Battery capacity   8 % - $ SmartUPS V1.00,Vin NG,BATCAP 8,Vout 5136 $
[2020-04-29 00:34:50] Battery capacity  81 % - $ SmartUPS V1.00,Vin NG,BATCAP 81,Vout 5136 $
[2020-04-29 00:35:04] Battery capacity  80 % - $ SmartUPS V1.00,Vin NG,BATCAP 80,Vout 5136 $


[2020-04-29 01:07:49] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5146 $
[2020-04-29 02:45:51] Battery capacity  10 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 10,Vout 5144 $
[2020-04-29 02:45:52] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 04:04:52] Battery capacity   0 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144 $
[2020-04-29 04:04:53] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 06:59:24] Battery capacity  10 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 10,Vout 5144 $
[2020-04-29 06:59:25] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $

[2020-04-29 10:02:31] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 10:23:57] Battery capacity   0 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144 $
[2020-04-29 10:23:58] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5143 $
[2020-04-29 10:36:57] Battery capacity   0 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144 $
[2020-04-29 10:36:58] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $
[2020-04-29 15:20:31] Battery capacity  10 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 10,Vout 5144 $
[2020-04-29 15:20:32] Battery capacity 100 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 100,Vout 5144 $

[2020-04-30 15:46:52] Battery capacity  72 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 72,Vout 5138 $
[2020-04-30 15:52:39] Battery capacity   7 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 7,Vout 5138 $
[2020-04-30 15:52:40] Battery capacity  72 % - $ SmartUPS V1.00,Vin GOOD,BATCAP 72,Vout 5138 $


As you can see, in a line like SmartUPS V1.00,Vin GOOD,BATCAP 00,Vout 5144  there is clearly a character cut off from the BATCAP value.
This might be a bug in the firmware on the Nuvoton N76E003 AT20 microcontroller. Can you provide a fixed firmware for flashing or maybe the source code for the board, so I can fix it myself?
Thank you!

**Comment by ProtoFoo** [comment]
Thanks for your suggestion. Checking the battery level for jumps could help with the problem.

But wouldn't you agree that the UPS board should not send wrong values in the first place?

**Comment by ashbashmash** [comment]
Absolutely. I know when I was trying to get any kind of decent reading though via a pi and c++ it was a bit hit and miss though. A lot of the time I’d get a serial read which didn’t even give an entire line, but I didn’t notice any major measurement mismatches….having said that I didn’t measure it for too long. I went with just trying to use what the board was giving me rather than looking into why the values weren’t correct in the first place. I don’t have the know how to dig into how the board provides its values, so just have to accept what it gives me and try and tweak the code to deal with the odd dodgy value. It does seem pretty clear the board is giving rubbish data occasionally, and the correct fix is to fix the source of the issue if you know how.

**Comment by ProtoFoo** [comment]
Yeah, I noticed in the Python code that the serial input is read in single characters. Which sometimes cuts the UPS data string in a weird way. I changed the code to read the data line by line with `serial.readline()` and by setting a timeout during setup with `serial.Serial(port, 9600, timeout=2)`. That works pretty solid.

But are you saying that you have no access to the actual code running on the Nuvoton microcontroller of the board? So there is no way to create a firmware update? That would be a shame because I really think it is a nice piece of hardware.

**Comment by ashbashmash** [comment]
I'm just a 3rd party consumer like you who happened to contribute some code in the hope it helped someone else, I have no connection to the makers of the board or the chip :( hence why I'm going for a workaround rather than attempting to fix the firmwareOn 5 May 2020 19:26, ProtoFoo <notifications@github.com> wrote:
Yeah, I noticed in the Python code that the serial input is read in single characters. Which sometimes cuts the UPS data string in a weird way. I changed the code to read the data line by line with serial.readline() and by setting a timeout during setup with serial.Serial(port, 9600, timeout=2). That works pretty solid.
But are you saying that you have no access to the actual code running on the Nuvoton microcontroller of the board? So there is no way to create a firmware update? That would be a shame because I really think it is a nice piece of hardware.

**Comment by ProtoFoo** [comment]
Well, okay then! Thanks for responding to my issue! Let's hope someone with access to the firmware responds.
